### PR TITLE
Add PDL pre-processing pass to compiler

### PR DESCRIPTION
**Pull request by nirvedhmeshram**
Allows a pdl script to be passed to iree-compile to be run in the pre-processing stage.

### CODE REVIEW

File: compiler/src/iree/compiler/Preprocessing/Common/test/preprocessing_pdl_patterns.mlir [hunk 1]
```diff
@@ -0,0 +1,98 @@
+// RUN: iree-compile --iree-preprocessing-pdl-spec-filename=%p/linalg.pdl.mlir %s --compile-to=preprocessing | FileCheck %s
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Preprocessing/Common/test/preprocessing_pdl_patterns.mlir]: This is fairly involved for a lit test. Could this be kept out of tree? Or maybe moved to a sample?

Comment (nirvedhmeshram) [reply compiler/src/iree/compiler/Preprocessing/Common/test/preprocessing_pdl_patterns.mlir]: Sure, we already have this out of tree, any thoughts on what I should show in the in-tree test then?

Comment (qedawkins) [reply compiler/src/iree/compiler/Preprocessing/Common/test/preprocessing_pdl_patterns.mlir]: For a lit test like this, any kind of proof of life is good. Like casting an i32 to index or something arbitrary, basically just to show that the patterns are being applied.

Comment (nirvedhmeshram) [reply compiler/src/iree/compiler/Preprocessing/Common/test/preprocessing_pdl_patterns.mlir]: I think I will move the test to samples, I am trying to show that the pass can be called with iree-compile so the transform has to be something that will lower e2e correctly. We already have tests for the pass itself.

File: compiler/src/iree/compiler/Preprocessing/Common/test/preprocessing_pdl_patterns.mlir [hunk 2]
```diff
@@ -0,0 +1,98 @@
+// RUN: iree-compile --iree-preprocessing-pdl-spec-filename=%p/linalg_pdl.mlir %s --compile-to=preprocessing | FileCheck %s
```
Comment (kuhar) [inline compiler/src/iree/compiler/Preprocessing/Common/test/preprocessing_pdl_patterns.mlir]: +1 to what @qedawkins said, I'd appreciate a test with a minimal example

File: compiler/src/iree/compiler/Pipelines/Options.cpp [hunk 3]
```diff
@@ -226,6 +226,11 @@ void PreprocessingOptions::bindOptions(OptionsBinder &binder) {
       llvm::cl::desc(
           "File name of a transform dialect spec to use for preprocessing"),
       llvm::cl::cat(category));
+  binder.opt<std::string>(
+      "iree-preprocessing-pdl-spec-filename", preprocessingPDLSpecFilename,
+      llvm::cl::desc(
+          "File name of a transform dialect spec to use for preprocessing"),
```
Comment (kuhar) [inline compiler/src/iree/compiler/Pipelines/Options.cpp:240]: We should document if both TD and PDL preprocessing can be specified at the same time and which one happens first

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
You can just add this to the MLP sample directory? That way you can have a lit test and you can have an e2e test?

File: compiler/src/iree/compiler/Preprocessing/Common/test/preprocessing_pdl_patterns.mlir (hunk 1 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Preprocessing/Common/test/preprocessing_pdl_patterns.mlir]: Hmm, I dont understand the issue here...  this lit tests seems pretty simple to me.

File: samples/pdl/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,34 @@
+# Copyright 2023 The IREE Authors
```
Comment (kuhar) [inline samples/pdl/BUILD.bazel]: nit
```suggestion
# Copyright 2024 The IREE Authors
```

File: compiler/src/iree/compiler/Preprocessing/Common/test/preprocessing_pdl_patterns.mlir (hunk 1 above)
Comment (nirvedhmeshram) [reply compiler/src/iree/compiler/Preprocessing/Common/test/preprocessing_pdl_patterns.mlir]: I added a small test here and the other test in mlp samples.

File: samples/custom_dispatch/cpu/mlp_plugin/mlp_torch.mlir [hunk 5]
```diff
@@ -44,10 +44,6 @@
 
 #map = affine_map<(d0, d1) -> (d0, d1)>
 module @example attributes {hal.device.targets = [#cpu_target]} {
-
-  // CHECK-LABEL: EXEC @mlp_invocation
-  //       CHECK: [Plugin]: M = 2, N = 8, K = 4
-  //       CHECK: 2x8xf32=[-24 -0 -24 -0 -24 -0 -24 -0][-0 -24 -0 -24 -0 -24 -0 -24]
```
Comment (nirvedhmeshram) [inline samples/custom_dispatch/cpu/mlp_plugin/mlp_torch.mlir:50]: already present at the end of file.

**Comment by nirvedhmeshram** [comment]
@MaheshRavishankar I noticed a couple of missing filechecks in the other MLP tests so I added them, just want to make sure they werent omitted for a reason (windows failure or something of that effect).

File: compiler/src/iree/compiler/Preprocessing/Common/test/pdl_patterns.mlir [hunk 6]
```diff
@@ -0,0 +1,25 @@
+pdl.pattern @aith_optimization : benefit(1) {
+
+  %i64_type = pdl.type : f32
```
Comment (kuhar) [inline compiler/src/iree/compiler/Preprocessing/Common/test/pdl_patterns.mlir]:
```suggestion
  %f32_type = pdl.type : f32
```

File: compiler/src/iree/compiler/Preprocessing/Common/test/pdl_patterns.mlir [hunk 7]
```diff
@@ -0,0 +1,25 @@
+pdl.pattern @aith_optimization : benefit(1) {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Preprocessing/Common/test/pdl_patterns.mlir]:
```suggestion
pdl.pattern @arith_optimization : benefit(1) {
```

File: samples/custom_dispatch/cpu/mlp_plugin/mlp_linalg.mlir [hunk 8]
```diff
@@ -0,0 +1,109 @@
+// RUN: iree-compile --iree-preprocessing-pdl-spec-filename=%p/mlp_linalg_spec.pdl.mlir %s --compile-to=preprocessing | FileCheck %s
+
+// RUN: iree-compile --iree-preprocessing-pdl-spec-filename=%p/mlp_linalg_spec.pdl.mlir %s | \
+// RUN: iree-run-module --device=local-sync \
```
Comment (kuhar) [inline samples/custom_dispatch/cpu/mlp_plugin/mlp_linalg.mlir]:
```suggestion
// RUN:   iree-run-module --device=local-sync \
```

File: samples/custom_dispatch/cpu/mlp_plugin/mlp_linalg.mlir [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+// RUN:     --executable_plugin=$IREE_BINARY_DIR/samples/custom_dispatch/cpu/mlp_plugin/mlp_plugin$IREE_DYLIB_EXT \
+// RUN:     --module=- \
+// RUN:     --function=mlp_invocation \
+// RUN:     --input="2x4xf32=[[2.0, 2.0, 2.0, 2.0], [-2.0, -2.0, -2.0, -2.0]]" \
+// RUN:     --input="4x8xf32=[[3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0]]" | FileCheck %s --check-prefix=OUTPUT
```
Comment (kuhar) [inline samples/custom_dispatch/cpu/mlp_plugin/mlp_linalg.mlir]: This line is very long, can you put FileCheck on a new line?

File: samples/custom_dispatch/cpu/mlp_plugin/mlp_torch.mlir [hunk 10]
```diff
@@ -5,7 +5,7 @@
 // RUN:     --module=- \
 // RUN:     --function=mlp_invocation \
 // RUN:     --input="2x4xf32=[[2.0, 2.0, 2.0, 2.0], [-2.0, -2.0, -2.0, -2.0]]" \
-// RUN:     --input="4x8xf32=[[3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0]]"
+// RUN:     --input="4x8xf32=[[3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0]]" | FileCheck %s
```
Comment (kuhar) [inline samples/custom_dispatch/cpu/mlp_plugin/mlp_torch.mlir]: also here

File: samples/custom_dispatch/cpu/mlp_plugin/mlp_tosa.mlir [hunk 11]
```diff
@@ -5,7 +5,7 @@
 // RUN:     --module=- \
 // RUN:     --function=mlp_invocation \
 // RUN:     --input="2x4xf32=[[2.0, 2.0, 2.0, 2.0], [-2.0, -2.0, -2.0, -2.0]]" \
-// RUN:     --input="4x8xf32=[[3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0]]"
+// RUN:     --input="4x8xf32=[[3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0], [3.0, -3.0, 3.0, -3.0]]" | FileCheck %s
```
Comment (kuhar) [inline samples/custom_dispatch/cpu/mlp_plugin/mlp_tosa.mlir]: also here

File: compiler/src/iree/compiler/Pipelines/Options.cpp [hunk 12]
```diff
@@ -220,12 +220,23 @@ void PreprocessingOptions::bindOptions(OptionsBinder &binder) {
       llvm::cl::desc("Textual description of the pass pipeline to run before "
                      "running normal IREE compilation pipelines"),
       llvm::cl::cat(category));
+  // Two ways of doing custom transformations at the pre-processing step are
+  // supported.
+  // 1. Through a Transform dialect spec file.
```
Comment (kuhar) [inline compiler/src/iree/compiler/Pipelines/Options.cpp:227]: 0. Through a preprocessing pass pipeline?

**Review by kuhar — APPROVED** [review]
LGTM % nits

**Review by kuhar — APPROVED** [review]

File: samples/custom_dispatch/cpu/mlp_plugin/mlp.mlir [hunk 13]
```diff
@@ -4,7 +4,8 @@
 // RUN:     --module=- \
 // RUN:     --function=mlp_invocation \
 // RUN:     --input="2x2xf32=[[2.0, 2.0], [-2.0, -2.0]]" \
-// RUN:     --input="2x2xf32=[[3.0, -3.0], [3.0, -3.0]]"
+// RUN:     --input="2x2xf32=[[3.0, -3.0], [3.0, -3.0]]" | \
+// RUN: FileCheck %s
```
Comment (MaheshRavishankar) [inline samples/custom_dispatch/cpu/mlp_plugin/mlp.mlir:8]: Good catch! Thanks!

**Review by MaheshRavishankar — APPROVED** [review]

**Comment by ScottTodd** [comment]
`iree_tfl_tests/llvmcpu_resnet_50_int8.run` failed on this PR: https://github.com/openxla/iree/actions/runs/8543560157/job/23408563732#step:6:333

pkgci is non-blocking, so it allowed you to merge anyways. We might want to revert though.

**Comment by hanhanW** [comment]
Here is the stack trace: https://gist.githubusercontent.com/hanhanW/dfe96ae4e945d1d63b4e4d3abae6606a/raw/9b472a4fcfd16b68a88e518044b8818a632da806/log10

To reproduce it:

- Download the tflite file: https://storage.googleapis.com/tf_model_garden/vision/resnet50_imagenet/resnet_50_224_int8.tflite
- Run iree-import-tflite to convert it to mlir.
- Run `iree-compile --output-format=vm-bytecode --iree-hal-target-backends=llvm-cpu --iree-llvmcpu-target-cpu=cascadelake --iree-llvmcpu-target-triple=x86_64-unknown-linux-gnu ~/resnet_50_224_int8.mlirbc -o /tmp/z.vmfb`

I think it's better to revert the change, and re-land it later. Otherwise, there could be more people like me looking at this.

**Pull request merged**